usersession/agent/rest_api.go

```diff
@@ -46,3 +59,129 @@ func sessionInfo(c *Command, r *http.Request) Response {
+		time.Sleep(killWait)
+		sysd.Kill(service, "KILL", "")
+	}
+	return err
+}
+
+func serviceStart(inst *serviceInstruction, sysd systemd.Systemd) Response {
+	// Refuse to start non-snap services
+	for _, service := range inst.Services {
+		if !strings.HasPrefix(service, "snap.") {
+			return InternalError("cannot start service %v", service)
+		}
+	}
+
+	startErrors := make(map[string]string)
+	var started []string
+	for _, service := range inst.Services {
+		if err := sysd.Start(service); err != nil {
+			startErrors[service] = err.Error()
+			break
+		}
+		started = append(started, service)
+	}
+	// If we got any failures, attempt to stop the services we started.
+	if len(startErrors) != 0 {
+		for _, service := range started {
+			if err := stopOneService(service, sysd); err != nil {
+				startErrors[service] = err.Error()
+			}
+		}
+	}
+	if len(startErrors) == 0 {
+		return SyncResponse(nil)
+	}
+	return SyncResponse(&resp{
+		Type:   ResponseTypeError,
+		Status: 500,
+		Result: &errorResult{
+			Message: "some services failed to start",
+			Kind:    errorKindServiceControl,
+			Value:   startErrors,
+		},
+	})
+}
+
+func serviceStop(inst *serviceInstruction, sysd systemd.Systemd) Response {
+	// Refuse to start non-snap services
+	for _, service := range inst.Services {
+		if !strings.HasPrefix(service, "snap.") {
+			return InternalError("cannot stop service %v", service)
+		}
+	}
+
+	stopErrors := make(map[string]string)
+	for _, service := range inst.Services {
+		if err := stopOneService(service, sysd); err != nil {
+			stopErrors[service] = err.Error()
+		}
+	}
+	if len(stopErrors) == 0 {
+		return SyncResponse(nil)
+	}
+	return SyncResponse(&resp{
+		Type:   ResponseTypeError,
+		Status: 500,
+		Result: &errorResult{
+			Message: "some services failed to stop",
+			Kind:    errorKindServiceControl,
+			Value:   stopErrors,
+		},
+	})
+}
+
+func serviceDaemonReload(inst *serviceInstruction, sysd systemd.Systemd) Response {
+	if len(inst.Services) != 0 {
+		return InternalError("daemon-reload should not be called with any services")
+	}
+	if err := sysd.DaemonReload(); err != nil {
+		return InternalError("cannot reload daemon: %v", err)
+	}
+	return SyncResponse(nil)
+}
+
+var serviceInstructionDispTable = map[string]func(*serviceInstruction, systemd.Systemd) Response{
+	"start":         serviceStart,
+	"stop":          serviceStop,
+	"daemon-reload": serviceDaemonReload,
+}
+
+type dummyReporter struct{}
+
+func (dummyReporter) Notify(string) {}
+
+func postServices(c *Command, r *http.Request) Response {
+	if contentType := r.Header.Get("Content-Type"); contentType != "application/json" {
+		return BadRequest("unknown content type: %s", contentType)
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	var inst serviceInstruction
+	if err := decoder.Decode(&inst); err != nil {
+		return BadRequest("cannot decode request body into service instruction: %v", err)
+	}
+	impl := serviceInstructionDispTable[inst.Action]
```

Comment:
Do we need to grab a lock for this? If several instructions are posted concurrently I think we should lock around either globally (per user) or per snap (and refuse to issue commands that would affect multiple snaps in one go).

Reply:
That seems reasonable.  I've added a global lock to serialise the final final portion of `postServices`.

Reply:
@zyga - I believe @jhenstridge change should address your concerns?